src/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/Base64Transforms.cs

```diff
@@ -96,135 +103,217 @@ public class FromBase64Transform : ICryptoTransform
+            if (bytesToTransform < InputBlockSize)
             {
+                // reinitialize the transform
                 Reset();
+
                 return Array.Empty<byte>();
             }
 
-            byte[] result = ConvertFromBase64(temp, effectiveCount);
+            int outputSize = GetOutputSize(bytesToTransform, tmpBuffer);
+            byte[] output = new byte[outputSize];
+
+            ConvertFromBase64(tmpBuffer, output, out int consumed, out int written);
+            Debug.Assert(consumed == bytesToTransform);
+            Debug.Assert(written == outputSize);
+
+            if (tmpBufferArray != null)
+            {
+                tmpBuffer.Clear();
+                ArrayPool<byte>.Shared.Return(tmpBufferArray);
+            }
 
             // reinitialize the transform
             Reset();
 
-            return result;
+            return output;
         }
 
-        private byte[] GetTempBuffer(byte[] inputBuffer, int inputOffset, int inputCount, out int effectiveCount)
+        private Span<byte> GetTempBuffer(Span<byte> inputBuffer, Span<byte> tmpBuffer)
         {
-            byte[] temp;
-
-            if (_whitespaces == FromBase64TransformMode.IgnoreWhiteSpaces)
+            if (_whitespaces == FromBase64TransformMode.DoNotIgnoreWhiteSpaces)
             {
-                temp = DiscardWhiteSpaces(inputBuffer, inputOffset, inputCount);
-                effectiveCount = temp.Length;
+                return inputBuffer;
             }
-            else
+
+            return DiscardWhiteSpaces(inputBuffer, tmpBuffer);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Span<byte> DiscardWhiteSpaces(Span<byte> inputBuffer, Span<byte> tmpBuffer)
+        {
+            int count = 0;
+
+            for (int i = 0; i < inputBuffer.Length; i++)
             {
-                temp = new byte[inputCount];
-                Buffer.BlockCopy(inputBuffer, inputOffset, temp, 0, inputCount);
-                effectiveCount = inputCount;
+                if (!IsWhitespace(inputBuffer[i]))
+                {
+                    tmpBuffer[count++] = inputBuffer[i];
+                }
             }
 
-            return temp;
+            return tmpBuffer.Slice(0, count);
+        }
+
+        private static bool IsWhitespace(byte value)
+        {
+            // We assume ASCII encoded data. If there is any non-ASCII char, so it is invalid
+            // Base64 and will be caught during decoding.
+
+            // SPACE        32
+            // TAB           9
+            // LF           10
+            // VTAB         11
+            // FORM FEED    12
+            // CR           13
+
+            // https://github.com/dotnet/coreclr/issues/914
+            return value == 32 || ((uint)value - 9 <= (13 - 9)) ? true : false;
         }
 
-        private byte[] ConvertFromBase64(byte[] temp, int effectiveCount)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
I'm not a fan of forcing the compiler's hand and doing so can cause output code bloat. Is this needed?

Reply:
Here it results in more streamlined code, and total code-size is less than without forcing the inline.
Without there is a lot stack-work done, so this can be avoided.

Reply:
Is there an issue tracking whatever improvements to the JIT would enable the removal of this AggressiveInlining?
cc: @AndyAyersMS

Reply:
There is just one caller, so probably not too much worry about bloat. But I'd be surprised if forcibly inlining this measurably impacted perf.

Reply:
> if forcibly inlining this measurably impacted perf

```
|                     Method |      Mean |     Error |    StdDev | Ratio | RatioSD |
|--------------------------- |----------:|----------:|----------:|------:|--------:|
|                         PR |  97.83 ns | 1.9625 ns | 2.0999 ns |  1.00 |    0.00 |
| PR_woInliningGetOutputSize | 106.60 ns | 0.8049 ns | 0.7529 ns |  1.09 |    0.03 |
```
The second benchmark is without forcing inling. The resulting asm is just 64 bytes less.
The inlined version has really nice asm, that's why I forced it.